server/app/views/admin/programs/ProgramBlockPredicateConfigureView.java

```diff
@@ -0,0 +1,621 @@
+    DivTag container =
+        div("Screen is ").withClasses("font-bold", ReferenceClasses.PREDICATE_ACTION);
+    SelectTag selectTag = select().withName("predicateAction");
+
+    ImmutableList.of(PredicateAction.HIDE_BLOCK, PredicateAction.SHOW_BLOCK)
+        .forEach(
+            action ->
+                selectTag.with(
+                    option(action.toDisplayString())
+                        .withClasses(
+                            ReferenceClasses.MULTI_OPTION_QUESTION_OPTION,
+                            ReferenceClasses.MULTI_OPTION_VALUE)
+                        .withCondSelected(
+                            maybeExistingPredicate
+                                .map(PredicateDefinition::action)
+                                .map(existingAction -> existingAction.equals(action))
+                                .orElse(false))
+                        .withValue(action.name())));
+
+    container.with(selectTag);
+
+    return container;
+  }
+
+  private ImmutableList<PredicateExpressionNode> getExistingAndNodes(
+      PredicateDefinition existingPredicate) {
+    switch (existingPredicate.computePredicateFormat()) {
+      case SINGLE_QUESTION:
+        {
+          return ImmutableList.of(
+              PredicateExpressionNode.create(
+                  AndNode.create(ImmutableList.of(existingPredicate.rootNode()))));
+        }
+
+      case OR_OF_SINGLE_LAYER_ANDS:
+        {
+          return existingPredicate.rootNode().getOrNode().children();
+        }
+
+      default:
+        {
+          throw new RuntimeException(
+              String.format(
+                  "Unrecognized predicate format: %s", existingPredicate.computePredicateFormat()));
+        }
+    }
+  }
+
+  private DivTag renderQuestionHeaders(
+      ImmutableList<QuestionDefinition> questionDefinitions,
+      Optional<PredicateDefinition> maybeExistingPredicate) {
+    DivTag container = div().withClasses("flex", "py-4");
+    int columnNumber = 1;
+
+    if (maybeExistingPredicate.isPresent()) {
+      ImmutableList<LeafOperationExpressionNode> leafNodes =
+          getExistingAndNodes(maybeExistingPredicate.get()).stream()
+              .findFirst()
+              .get()
+              .getAndNode()
+              .children()
+              .stream()
+              .map(PredicateExpressionNode::getLeafNode)
+              .sorted(
+                  (LeafOperationExpressionNode loenA, LeafOperationExpressionNode loenB) ->
+                      (int) (loenA.questionId() - loenB.questionId()))
+              .collect(ImmutableList.toImmutableList());
+
+      for (LeafOperationExpressionNode leafNode : leafNodes) {
+        long questionId = leafNode.questionId();
+        QuestionDefinition qd =
+            questionDefinitions.stream()
+                .filter(questionDefinition -> questionDefinition.getId() == questionId)
+                .findFirst()
+                .get();
```

Comment:
need to handle findFirst returning an empty optional and providing a useful error message.  

This could occur if our versioning updating has a bug in it........

Reply:
I don't think it could since the controller filters the questions based on what's in the predicate.

The logic is different since my latest push anyways.